### ISSUE TITLE
Prettier is moving ts-ignore comment

### ISSUE THREAD

**Issue by michaeljota** [prettier/prettier#5411]
**Prettier 1.15.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEcAeAHCAnGACMAGwEMBnUvAdTgCMNiwBrAFW2KlJPgBkJiATONjwBLALYZCcMQhgVW7TsR59Bw4AB0oePBgCuNQiLB4A5nBgKOXEdAA8zAHwAKElFNI8pGNhHuAlJ4A8jSkQgBuxIZwDo54mto6eNgWetjaAGbYEGLOWklJAPSFeAACcgC0IqZQOHB4+QWiEjgwzgDk-H7EYsYAshD8elLt-o06-gDcjQC+WnNQIAA0IBAYMLYcyKDE2NkA7gAKuwikyCDE4RAi-MsgNGxMFgDK9GB+psg+enArABYwMSESh-ETwUhvODPU5gkThMEAT3OYHIdz8YVwhzYpjExGQGWIhDCKwAVqQ0AAhR6MF49ODcPxwfGE4kgMloZ4fKQART0EHgzKJvxA9GwGPOMARGDgpDAvnWdwwvlglBuMD+yAAHAAGFZKiBhShsDDnJUyiJMlYpACOehEKSxxBxeKQBKFKzCvS+2B+Hq5cF5-KZrpZwpgUVV-HVyAATCsfMQREZ3ABhHK4861KCWkB6MLMKJnENCmYzIA)
```sh
# Options (if any):
--single-quote
```

**Input:**
```ts
export class WebpackTranslateLoader implements TranslateLoader {
  public getTranslation<T>(lang: string): Observable<T> {
    return from(
      // @ts-ignore 
      import('dinamicModule')
    );
  }
}
```

**Output:**
```ts
export class WebpackTranslateLoader implements TranslateLoader {
  public getTranslation<T>(lang: string): Observable<T> {
    return from(
      import(// @ts-ignore
      "dinamicModule")
    );
  }
}
```

**Expected behavior:**

Same as input. The comment is mean to ignore the function call, but by moving the comment inside the function, the meaning of the comment is lost.

**Comment by Loilo** [comment]
This is a really hard problem, not just a "move `@ts-ignore` elsewhere" issue.

`@ts-ignore` ignores one line of code, not one expression or statement (a behavior diametrically opposed to code beautifiers which operate on the AST). It therefore almost always affects code where Prettier condenses or spreads out TypeScript code.

### Line Spreading
Below is an example from my current project (which brought me to this issue). The `@ts-ignore` is prepended because `arg.node.value` throws an error due to wrong external typing.

This is the original code:

```ts
// @ts-ignore
const result = evaluate(chain, opts.filename, args.map(arg => arg.node.value))
```

It becomes:
```ts
// @ts-ignore
const result = evaluate(
  chain,
  opts.filename,
  args.map(arg => arg.node.value)
)
```

When it should become:
```ts
const result = evaluate(
  chain,
  opts.filename,
  // @ts-ignore
  args.map(arg => arg.node.value)
)
```

And since there's no way for Prettier to know which is the part to be ignored, it should either bail out of formatting that line completely or prepend a `@ts-ignore` before every single line:

```ts
// @ts-ignore
const result = evaluate(
  // @ts-ignore
  chain,
  // @ts-ignore
  opts.filename,
  // @ts-ignore
  args.map(arg => arg.node.value)
// @ts-ignore
)
```

This is pragmatic yet in no way... prettier than before. Therefore I'd vote for bailing out on a `@ts-ignore` encounter.

### Line Condensing

However, even bailing out is not always trivial. It becomes way harder when the line below the `@ts-ignore` is only part of an expression or block.

Take this example:
```ts
// @ts-ignore
const foo = () =>
  bar()
```

which is prettified to:

```ts
// @ts-ignore
const foo = () => bar();
```

It's not strictly breaking anything, but it might in rare edge cases suppress TypeScript errors that were not supposed to be suppressed.

---

No actual bottom line here, those are just some cases to consider when attempting to fix this issue.

**Comment by michaeljota** [comment]
@Loilo 

> This is the original code:
> ```ts
> // @ts-ignore
> const result = evaluate(chain, opts.filename, args.map(arg => arg.node.value))
> ```
> It becomes:
> ```ts
> // @ts-ignore
> const result = evaluate(
>   chain,
>   opts.filename,
>   args.map(arg => arg.node.value)
> )
> ```
> When it should become:
> ```ts
> const result = evaluate(
>   chain,
>   opts.filename,
>   // @ts-ignore
>   args.map(arg => arg.node.value)
> )
> ```

I don't think this is true. Yes, the formatter would break the statement, but it should be manually fixed as there is no way for Prettier to understand that this is the desired behavior. 

The issue is actually about that, instead of trying to understand what the `// @ts-ignore` is doing, just _ignore_ it, and format everything around it. If, as you expose, it breaks something, I think is better that it does it by omission, than because Prettier actually tried to understand where the comment belongs.

**Comment by Loilo** [comment]
That sounds reasonable. I actually considered writing down that option too but apparently somehow forgot about that, sorry.

It's probably the best approach as 100% correct behaviour is probably unachievable with either solution applied, so the one wth the least disruption should be chosen.

**Comment by lvl99** [comment]
I'm gonna throw in my two cents on a `// @ts-ignore` use-case I'm trying to use, which prettier seems to break:

```typescript
function flattenOptions(
  options: ReactSelectOption[]
): ReactSelectOptionObject[] {
  return options.reduce(
    (acc: ReactSelectOptionObject[], option: ReactSelectOption) =>
      option.hasOwnProperty("options")
        // @ts-ignore: totally legal thing to do
        ? flattenOptions(option.options)
        // @ts-ignore: another totally legal thing to do
        : acc.concat([option]),
    []
  );
}
```

gets formatted like so and means the `// @ts-ignore` lines aren't in effect:

```typescript
function flattenOptions(
  options: ReactSelectOption[]
): ReactSelectOptionObject[] {
  return options.reduce(
    (acc: ReactSelectOptionObject[], option: ReactSelectOption) =>
      option.hasOwnProperty("options")
        ? // @ts-ignore: totally legal thing to do
          flattenOptions(option.options)
        : // @ts-ignore: another totally legal thing to do
          acc.concat([option]),
    []
  );
}
```

**Comment by j-f1** [comment]
Does the formatted version no longer ignore the issues?

**Comment by lvl99** [comment]
@j-f1 yeah, it stops the effectiveness of the `// @ts-ignore`. I think `// @ts-ignore` lines have to occupy a single line alone. It seems only whitespace is allowed before the comment on the same line?

**Comment by lydell** [comment]
As a workaround, this might work:

**Prettier 1.16.4**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzArlMMCW0AEqANgIYzxQDyADrtAM4AUAOlPvhLXlPUvgEpwS2AMpwicbDTpQA2gF1WASj6DhMMRKldolAEYArSTAX5grdgCc4MdJbacZ9AHTWAJujBwWbdvkbCYKpCouLG0tz6RtgKADQcOlDB6prhiUr4ALwAfBZ+7I7czgAWJPSUAO5QAAqWnHCWMACeLCCFDMwgSnn57AD0ffgeALbDTfiQowgw+DAQ+MMkANZws-QAtDgA5lAQ1vgVe0s9vQP4AAIwG9u71nxzMCREROMSW0+zxThQW7PzbhATvkAPyEUjkBARBiMdpQZyw+jdXy9fBnEZjCYQKawP4LZarK6bHZ7VaHSzHZGnQaXa7Eu74Ei7GDFBp-R7PV5wd5ET7fX5zIaAyn5PiBZyQLBkRiyWHyJSxE4KPJKADcrAAviBYm1EvRkKASJY6hVqoaEHqUCQAG4QHBuLUgPSWYQrDTUYR85AwSzoODa4owYZEADqX3g9HdXjEPBwuCtsaayHA9D12u+9AaMFqJC2i2QqCeGe1BnoAA8AELOsCukQkYZwAAy3zg+cLfpAJdLIj5EgAiugIPBW0QiyB3ZYM5Yk3oSHomkRoA7qJZvjBg3bmcgABwABm1y4gGeDzuoSeXcEnVpb2usAEd0DhrNncyRh6OM8McF6fe36D24P2g4tkgBYju2jx6OubibkgABM2reiQOBEHyADCWJ5iguxQNeIDoBmAAqs4WqBGbquqQA)
```sh
--parser babylon
```

**Input:**
```jsx
function flattenOptions(
  options: ReactSelectOption[]
): ReactSelectOptionObject[] {
  return options.reduce(
    (acc: ReactSelectOptionObject[], option: ReactSelectOption) =>
      option.hasOwnProperty("options")
        // dummy comment to make ts-ignore work
        // @ts-ignore: totally legal thing to do
        ? flattenOptions(option.options)
        // dummy comment to make ts-ignore work
        // @ts-ignore: another totally legal thing to do
        : acc.concat([option]),
    []
  );
}
```

**Output:**
```jsx
function flattenOptions(
  options: ReactSelectOption[]
): ReactSelectOptionObject[] {
  return options.reduce(
    (acc: ReactSelectOptionObject[], option: ReactSelectOption) =>
      option.hasOwnProperty("options")
        ? // dummy comment to make ts-ignore work
          // @ts-ignore: totally legal thing to do
          flattenOptions(option.options)
        : // dummy comment to make ts-ignore work
          // @ts-ignore: another totally legal thing to do
          acc.concat([option]),
    []
  );
}

```

**Comment by lvl99** [comment]
Sweet trick @lydell — it works! However a part of me wishes that this could be an internal rule that prettier can automatically work around.

**Comment by duailibe** [comment]
> I think `// @ts-ignore` lines have to occupy a single line alone

Wow that's unfortunate

**Comment by glen-84** [comment]
This seems to be fixed now?

**Comment by evilebottnawi** [comment]
Yes, original issue was solved, feel free to open new issue if you have other problems with comments, thanks

**Comment by thorn0** [comment]
The specific example from the top post works now, but the [general problem](https://github.com/prettier/prettier/issues/5411#issuecomment-445971247) with `// @ts-ignore` has not been solved. Let's reopen this issue as it has valuable comments.

**Comment by evilebottnawi** [comment]
@thorn0 I prefer new issue, https://github.com/prettier/prettier/issues/5411#issuecomment-445971247 should be solved manually not on prettier side, we can't duplicate comments, same problem with eslint comments and etc, we already discussed about it

**Comment by evilebottnawi** [comment]
Comments inside ternary have other issue, so we can close this issue

**Comment by thorn0** [comment]
I'm not talking only about ternaries. I mean the general problem with the fact that `ts-ignore` is applied on the line level. The solution here might be to treat it as `prettier-ignore` placed in front of the **whole statement**.

**Comment by evilebottnawi** [comment]
@thorn0 still can't understand problem and how we can solve this, please clarify

**Comment by thorn0** [comment]
The problem is described in the comment I linked to.

In order to not break anything,
```ts
foo(
  baz,
  // @ts-ignore
  bar
);
```
should be treated as
```ts
// prettier-ignore
foo(
  baz,
  // @ts-ignore
  bar
);
```

**Comment by thorn0** [comment]
Oops, I missed this part:

> It's not strictly breaking anything, but it might in rare edge cases suppress TypeScript errors that were not supposed to be suppressed.

So there seems to be no solution for the general case. Okay, let's close this issues as non-actionable.

**Comment by thorn0** [comment]
Sorry, I deleted my previous comment. It was completely wrong. Ignoring the whole statement seems to be a solution, albeit an ugly one. So this issue is fixable.

**Comment by loilo** [comment]
Also my statement which you cited in the deleted comment is completely bogus, gonna update my comment.

**Comment by evilebottnawi** [comment]
@thorn0 i disagree, we should not ignore nodes with comments, other developers can potential use other comments with any values and we can't change output based on comment text

**Comment by thorn0** [comment]
> we can't change output based on comment text

We already do so with JSDoc type assertions. And `// prettier-ignore` itself is a comment too.

**Comment by glen-84** [comment]
That solution would also cause other parts to be ignored, so:

```js
foo(
baz,
// @ts-ignore
bar,baz
);
```

... would not be formatted at all?

**Comment by thorn0** [comment]
@glen-84 Right. But what else can be done here?

**Comment by evilebottnawi** [comment]
> We already do so with JSDoc type assertions. And // prettier-ignore itself is a comment too.

@thorn0 please provide example

**Comment by evilebottnawi** [comment]
Also original issue was about other problem, if you want to discussion how we can solve post/feedback from the issue you need open an new issue, it is very hard to track problems inside thread in other issue

**Comment by evilebottnawi** [comment]
We also try to ignore original input as much as possible when we format, so you change formatting based on comment text is horrible idea for me

**Comment by thorn0** [comment]
@evilebottnawi Do you really need an example of how `// prettier-ignore` affects the output? :)

I wouldn't say the original issue was different. The title and the discussion of this issue are about the generic problem. I really don't see why open a new one.

Using `ts-ignore` is a bad idea too and should not be used, so I actually somewhat like the fact that Prettier breaks it. But still, imagine a situation where Prettier runs on the pre-commit hook: the code compiles before the commit, you commit, Prettier breaks ts-ignore, now it doesn't compile for everyone...

**Comment by evilebottnawi** [comment]
@thorn0 okay, please rename issue and move main idea of problem on top

**Comment by glen-84** [comment]
@thorn0,

The issue is not limited to `@ts-ignore` comments. If I place an end-of-line comment next to a line, I expect it to stay there (see also #807 and #7155).

There should perhaps be a single generic issue to solve the general problem of end-of-line comments being moved.

**Comment by thorn0** [comment]
@glen-84 

> The issue is not limited to @ts-ignore comments

**This** issue definitely is. Anything else is off topic here.

**Comment by glen-84** [comment]
My point is that solving the issue with moving comments solves this problem automatically, so a separate issue for this is not necessary.

**Comment by glen-84** [comment]
Let me make that more clear: I now have this issue with `banana-ignore` comments – what do I do?

**Comment by evilebottnawi** [comment]
@glen-84 right now - manually fix it, CI should be allow to catch any problems with lint/ts/etc

**Comment by thorn0** [comment]
@glen-84 Please see https://github.com/prettier/prettier/issues/5411#issuecomment-445971247 if you really want to understand this issue.

**Comment by glen-84** [comment]
@evilebottnawi It was hypothetical.

@thorn0 Apologies, I missed that spreading case, but I still believe that:

1. Ignoring formatting on the whole statement is far from ideal.
2. A similar issue could easily occur with other tools using other comments.

I don't see how this issue could be solved.

**Comment by thorn0** [comment]
`@ts-ignore` is supposed to be used as a temporary measure. It's very much like a `TODO` comment. So fix all `@ts-ignore`s and now all your code is formatted again. It really looks like a good enough trade-off.

**Comment by glen-84** [comment]
Would it be possible to disable the spreading behaviour for the line following a `ts-ignore` comment?

In that way, at least the rest of the line is formatted correctly.

**Comment by thorn0** [comment]
No, Prettier doesn't operate on lines.

**Comment by glen-84** [comment]
The next AST node then.

**Comment by loilo** [comment]
Wouldn't be sufficient either.

```ts
// @ts-ignore
const foo = 1; const bar = 2;
```

**Comment by glen-84** [comment]
Who writes code like that? 😛

Also, the previously-suggested solution would have the same limitation.

I don't know anything about the internals of Prettier, but perhaps you could use the location metadata in the AST to ignore all nodes on the line following the comment.

**Comment by thorn0** [comment]
@glen-84 It's not worth the trouble. `@ts-ignore` is like the `with` statement or the comma operator. We just don't want Prettier to break them. That's it. We shouldn't really care how they look in the end. That said, if you feel like trying to implement a more sophisticated solution, I'll be glad to review the PR.

**Comment by lobsterkatie** [comment]
Just pinging here, as I just ran into exactly this problem (from earlier in this thread) with ternaries and VSCode: https://github.com/prettier/prettier/issues/5411#issuecomment-461881755

The workaround with dummy comments does work, but it would be great if the underlying problem could be fixed.

NB: Someone mentioned that there is a separate issue for ternaries, but I couldn't find it. Happy to move this there if someone can point me in the right direction.

Thanks!

**Comment by ddubrava** [comment]
I know it's obvious, but just to mention, it's not only about `@ts-ignore`; the same thing goes for `eslint-disable` comments

<img width="1013" alt="Screenshot 2023-07-18 at 20 01 21" src="https://github.com/prettier/prettier/assets/22116465/f0097346-d8fd-41e9-930e-de223343ed35">

**Comment by so1ve** [comment]
IMO prettier should maintain a list of "directive comments", and ignore lines below these comments.

**Comment by iteriani** [comment]
fwiw, within google we've locally patched prettier to treat ts-ignore as prettier-ignore. This seems to work alright.

**Comment by lobsterkatie** [comment]
@iteriani - Neat idea. To save some digging, mind sharing how you did that?

**Comment by iteriani** [comment]
There's some function called

function isPrettierIgnoreComment(comment) {}

We basically did something like

      return (comment.value.trim() === "prettier-ignore" || comment.value.trim().includes("@ts-ignore")) && !comment.unignore;

**Comment by lobsterkatie** [comment]
Great, will give it a shot, thanks!

**Comment by lucasbasquerotto** [comment]
I just want to point out that it's advisable to use `@ts-expect-error` (e.g., `// @ts-expect-error TS2589`) instead of `@ts-ignore`. With `@ts-expect-error`, you can explicitly specify the error type you want to suppress. Additionally, if a future update (such as a dependency fix or a change in Prettier’s formatting) causes the error to disappear or shift to a different line, the linter will detect that no error is being suppressed anymore. In such cases, it will raise a warning or error at the `@ts-expect-error` comment, signaling that you should either remove or relocate it.

This helps address some of the problems caused by misplaced comments (such as suppressing unintended errors or failing to apply to the correct line). While this doesn't eliminate the issue entirely, it serves as a useful mitigation strategy until a proper solution is implemented. Regardless of this specific issue, `@ts-expect-error` is generally a better choice than `@ts-ignore`.

_(This isn’t specifically related to the issue of comments, but it can help mitigate some of the problems caused by it)_
